### PR TITLE
Port lookup and registration for MethodGenFn

### PR DESCRIPTION
**Pull request by noahgibbs**
Plus some method generation functions.

I moved rb_method_entry_t and rb_callable_method_entry_t into bindgen rather than wrapping all their accessors. They're pretty simple, so it seems like a win.

I added origin headers to the opaque structs in cruby.rs.

### CODE REVIEW

File: yjit/src/codegen.rs [hunk 1]
```diff
@@ -3326,35 +3327,37 @@ fn jit_protected_callee_ancestry_guard(jit: &mut JITState, cb: &mut CodeBlock, o
...
 {
-    ID mid = rb_intern(mid_str);
-    const rb_method_entry_t *me = rb_method_entry_at(klass, mid);
+    let id_string = std::ffi::CString::new(mid_str).expect("couldn't convert to CString!");
+    let mid = unsafe { rb_intern(id_string.as_ptr()) };
+    let me = unsafe { rb_method_entry_at(klass, mid )};
 
-    if (!me) {
-        rb_bug("undefined optimized method: %s", rb_id2name(mid));
+    if me == (0 as *const rb_method_entry_t) {
```
Comment (XrXr) [inline yjit/src/codegen.rs]:
```suggestion
    if me.is_null() {
```

Comment (noahgibbs) [reply yjit/src/codegen.rs]: Ah, okay. And there are a few places I can use that. Thanks!

File: yjit/src/codegen.rs [hunk 2]
```diff
@@ -3725,7 +3718,7 @@ fn gen_return_branch(cb: &mut CodeBlock, target0: CodePtr, target1: Option<CodeP
 
 fn gen_send_iseq(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut OutlinedCb, ci: * const rb_callinfo, cme: * const rb_callable_method_entry_t, block: Option<IseqPtr>, argc: i32) -> CodegenStatus
 {
-    let iseq = unsafe { def_iseq_ptr(get_cme_def(cme)) };
+    let iseq = unsafe { def_iseq_ptr((*cme).def) };
```
Comment (maximecb) [inline yjit/src/codegen.rs]: If this function is a getter, maybe it should be named `get_def_iseq_ptr` ?

File: yjit/src/codegen.rs [hunk 3]
```diff
@@ -5198,28 +5191,30 @@ fn get_method_gen_fn()
...
+        /*
+        // rb_str_to_s() methods in string.c
+        yjit_reg_method(rb_cString, "to_s", jit_rb_str_to_s);
+        yjit_reg_method(rb_cString, "to_str", jit_rb_str_to_s);
+        yjit_reg_method(rb_cString, "bytesize", jit_rb_str_bytesize);
+
+        // Thread.current
+        yjit_reg_method(rb_singleton_class(rb_cThread), "current", jit_thread_s_current);
+        */
+    }
```
Comment (maximecb) [inline yjit/src/codegen.rs:5217]: This doesn't seem like it needs to be wrapped in unsafe.

File: yjit/src/codegen.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: It's an older one from before that naming convention. Let me see if it's the last one left or if there are more.

File: yjit/src/codegen.rs (hunk 3 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs:5217]: Every class constant is a mutable static. If we don't wrap it in an unsafe block then every individual line needs to be wrapped in unsafe.

Comment (maximecb) [reply yjit/src/codegen.rs:5217]: Oh you mean rb_cSymbol, etc need to be wrapped in unsafe?

File: yjit/src/codegen.rs (hunk 2 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs]: Okay. I think that's the only one with that naming convention problem, though there are a few other inconsistencies of other types (e.g. functions in cruby.rs that could be moved over to bindgen.)

File: yjit/src/codegen.rs (hunk 3 above)
Comment (noahgibbs) [reply yjit/src/codegen.rs:5217]: That's what I mean. If you pull one of those lines out of the unsafe block and try to typecheck it locally, it should give you an error message telling you so.

**Pull request merged**